### PR TITLE
ODSC 39392/triton

### PR DESCRIPTION
**Pull request by z7ye**
## Description
https://jira.oci.oraclecorp.com/browse/ODSC-39392

## Testing Result
image

## Int Test Result
You can check this pdf for testing results below.
[inference-approach1.pdf](https://github.com/oracle/accelerated-data-science/files/11194180/inference-approach1.pdf)

### CODE REVIEW

File: ads/model/deployment/model_deployment.py [hunk 1]
```diff
@@ -1566,6 +1584,9 @@ def _build_model_deployment_configuration_details(self) -> Dict:
                 infrastructure.web_concurrency
             )
             runtime.set_spec(runtime.CONST_ENV, environment_variables)
+        if hasattr(runtime, "inference_server") and runtime.inference_server and runtime.inference_server.lower() == "triton":
```
Comment (mayoor) [inline ads/model/deployment/model_deployment.py:1587]: "triton" should be a constant

File: ads/model/deployment/model_deployment_runtime.py [hunk 2]
```diff
@@ -532,3 +534,54 @@ def with_health_check_port(
...
+        ...                 .with_predict_log(log_group_id=<deployment_log_group_id>, log_id=<deployment_predict_log_id>)
+
+        >>> runtime = ModelDeploymentContainerRuntime()\
+        ...                 .with_image(<container_image>)\
+        ...                 .with_server_port(<server_port>)\
+        ...                 .with_health_check_port(<health_check_port>)\
+        ...                 .with_model_uri(<model_id>)\
+        ...                 .with_env({"key":"value", "key2":"value2"})\
+        ...                 .with_inference_server("triton")
+        ...                 deployment = ModelDeployment()\
```
Comment (mayoor) [inline ads/model/deployment/model_deployment_runtime.py:581]: should there be `>>>` remove the extra indentation here?

File: ads/model/deployment/model_deployment.py [hunk 3]
```diff
@@ -916,9 +923,16 @@ def predict(
             raise TypeError(
                 "`data` is not bytes or json serializable. Set `auto_serialize_data` to `True` to serialize the input data."
             )
-
+        if model_name and model_version:
+            header['model-name'] = model_name
+            header['model-version'] = model_version
+        elif not model_version and not model_name:
```
[the remark this answers is not in the document]
Comment (z7ye) [reply ads/model/deployment/model_deployment.py:929]: Good call! fixed.

File: ads/model/deployment/model_deployment_runtime.py (hunk 2 above)
Comment (z7ye) [reply ads/model/deployment/model_deployment_runtime.py:581]: yes, good catch.

File: ads/model/deployment/model_deployment.py (hunk 1 above)
Comment (z7ye) [reply ads/model/deployment/model_deployment.py:1584]: fixed.

File: ads/model/deployment/common/utils.py [hunk 4]
```diff
@@ -140,7 +135,7 @@ def send_request(
             return json.loads(req.body)
         return req.body
     else:
-        request_kwargs["auth"] = header.get("signer")
+        request_kwargs["auth"] = header.pop("signer")
```
Comment (qiuosier) [inline ads/model/deployment/common/utils.py:138]: Shall we pop `signer` for dry_run as well?

File: ads/model/deployment/model_deployment.py (hunk 1 above)
Comment (qiuosier) [reply ads/model/deployment/model_deployment.py:1584]: We can use `upper()` instead then we can use the same constant (`MODEL_DEPLOYMENT_INFERENCE_SERVER_TRITON`) for the `TRITON` in the next line :)

File: ads/model/deployment/model_deployment.py [hunk 5]
```diff
@@ -1566,6 +1581,9 @@ def _build_model_deployment_configuration_details(self) -> Dict:
                 infrastructure.web_concurrency
             )
             runtime.set_spec(runtime.CONST_ENV, environment_variables)
+        if hasattr(runtime, "inference_server") and runtime.inference_server and runtime.inference_server.lower() == "triton":
+            environment_variables["CONTAINER_TYPE"] = "TRITON"
```
Comment (qiuosier) [inline ads/model/deployment/model_deployment.py:1585]:
```suggestion
            environment_variables["CONTAINER_TYPE"] = MODEL_DEPLOYMENT_INFERENCE_SERVER_TRITON
```

**Review by qiuosier — APPROVED** [review]

File: ads/model/deployment/model_deployment.py (hunk 1 above)
Comment (z7ye) [reply ads/model/deployment/model_deployment.py:1584]: fixed.

File: ads/model/deployment/common/utils.py (hunk 4 above)
Comment (z7ye) [reply ads/model/deployment/common/utils.py:138]: done

File: ads/model/deployment/model_deployment.py (hunk 5 above)
Comment (z7ye) [reply ads/model/deployment/model_deployment.py:1585]: fixed

**Review by lu-ohai — APPROVED** [review]

File: ads/model/deployment/model_deployment.py [hunk 6]
```diff
@@ -916,9 +923,13 @@ def predict(
             raise TypeError(
                 "`data` is not bytes or json serializable. Set `auto_serialize_data` to `True` to serialize the input data."
             )
-
+        if model_name and model_version:
+            header['model-name'] = model_name
+            header['model-version'] = model_version
+        elif bool(model_version) ^ bool(model_name):
```
Comment (mayoor) [inline ads/model/deployment/model_deployment.py:929]: We can rearrange this if/elif - 
```
if bool(model_version) ^ bool(model_name):
   raise ValueError("`model_name` and `model_version` have to be provided together.")
header['model-name'] = model_name
header['model-version'] = model_version
       
```

**Review by qiuosier — DISMISSED** [review]

**Review by qiuosier — APPROVED** [review]

**Pull request merged**